src/main/java/com/palantir/gradle/versions/CheckUnusedConstraintsTask.java

```diff
@@ -0,0 +1,163 @@
+import java.io.File;
+import java.io.IOException;
+import java.io.PrintWriter;
+import java.nio.file.Files;
+import java.nio.file.StandardOpenOption;
+import java.util.List;
+import java.util.Set;
+import java.util.stream.Collectors;
+import java.util.stream.Stream;
+import org.gradle.api.DefaultTask;
+import org.gradle.api.Project;
+import org.gradle.api.artifacts.Configuration;
+import org.gradle.api.artifacts.component.ModuleComponentIdentifier;
+import org.gradle.api.artifacts.result.ResolutionResult;
+import org.gradle.api.file.RegularFile;
+import org.gradle.api.file.RegularFileProperty;
+import org.gradle.api.provider.Property;
+import org.gradle.api.provider.Provider;
+import org.gradle.api.provider.SetProperty;
+import org.gradle.api.tasks.Input;
+import org.gradle.api.tasks.InputFile;
+import org.gradle.api.tasks.TaskAction;
+import org.gradle.api.tasks.options.Option;
+import org.gradle.language.base.plugins.LifecycleBasePlugin;
+
+public class CheckUnusedConstraintsTask extends DefaultTask {
+
+    private final Property<Boolean> shouldFix = getProject().getObjects().property(Boolean.class);
+    private final RegularFileProperty propsFileProperty = getProject().getObjects().fileProperty();
+    private final SetProperty<String> classpath = getProject().getObjects().setProperty(String.class);
+
+    public CheckUnusedConstraintsTask() {
+        shouldFix.set(false);
+        setGroup(LifecycleBasePlugin.VERIFICATION_GROUP);
+        setDescription("Ensures all versions in your versions.props correspond to an actual gradle dependency");
+    }
+
+    final void setPropsFile(File propsFile) {
+        this.propsFileProperty.set(propsFile);
+    }
+
+    @Input
+    public final SetProperty<String> getClasspath() {
+        return classpath;
+    }
+
+    @InputFile
+    public final Property<RegularFile> getPropsFile() {
+        return propsFileProperty;
+    }
+
+    @Option(option = "fix", description = "Whether to apply the suggested fix to versions.props")
+    public final void setShouldFix(boolean shouldFix) {
+        this.shouldFix.set(shouldFix);
+    }
+
+    final void setShouldFix(Provider<Boolean> shouldFix) {
+        this.shouldFix.set(shouldFix);
+    }
+
+    @TaskAction
+    public final void checkNoUnusedPin() {
+        Set<String> artifacts = getClasspath().get();
+        VersionsProps versionsProps = VersionsProps.loadFromFile(getPropsFile().get().getAsFile().toPath());
+
+        Set<String> exactConstraints = versionsProps.getFuzzyResolver().exactMatches();
+        Stream<String> unusedExactConstraints = Sets.difference(exactConstraints, artifacts).stream();
+
+        Sets.SetView<String> unmatchedArtifacts = Sets.difference(artifacts, exactConstraints);
+        Stream<String> unusedGlobConstraints = versionsProps.getFuzzyResolver().globs().stream()
+                .filter(glob -> !unmatchedArtifacts.stream().anyMatch(glob::matches))
+                .map(FuzzyPatternResolver.Glob::getRawPattern);
```

Comment:
the logic here should be more like
- start with all globs
- for each unmatchedArtifacts:
	- find the _most specific_ glob that matches
	- remove it from the set